### PR TITLE
geoip integrate air-gapped bootstrap script

### PR DESCRIPTION
**Pull request by kaisecheng**
enhancement

## Release notes

Bootstrap air-gapped environment for GeoIP database service

GeoIP plugin is expected to check update every day. For an air-gapped environment, Users can use a script, elasticsearch-geoip, to bootstrap a mock server to interact with Logstash. Set `xpack.geoip.download.endpoint` to the mock server.

## What does this PR do?

This PR integrates Elasticsearch bootstrap script to help user keep Logstah geoip plugin run without online update check.
Add `xpack.geoip.download.endpoint` option to config geoip database service endpoint. 
The default value is `https://geoip.elastic.co/v1/database`.
Users can point to `http://localhost:8080/overview.json` when using the [script](https://www.elastic.co/guide/en/elasticsearch/reference/master/geoip-processor.html#ingest-geoip-downloader-endpoint) to bootstrap nginx docker

## Why is it important/What is the impact to the user?

For security reasons, not all Logstash are free to access the internet. After 30 days without update, Maxmind EULA databases expire. The solution is setting up a mock server that mocks the interactions, a check update response and a download response.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files (and/or docker env variables)
- [x] I have added tests that prove my fix is effective or that my feature works

## Author's Checklist

- [ ] please follow the steps to test it locally
- [ ] please check the [doc](https://github.com/logstash-plugins/logstash-filter-geoip/pull/189) if it is clear 

## How to test this PR locally

- download latest elasticsearch
- copy $ES_HOME/modules/ingest-geoip/*.mmdb to YOUR_DB_DIR
- generate mock response,  `./bin/elasticsearch-geoip -s YOUR_DB_DIR`, `overview.json` file is generated
- in `logstash.yml`, set `xpack.geoip.download.endpoint: "http://localhost:8080/overview.json"`
- start mock server nginx, `docker run -d --name=nginx_geoip -p 8080:80 -v YOUR_DB_DIR:/usr/share/nginx/html:ro nginx`
- start Logstash with geoip. You should see no error in log and see database download in $LS_HOME/data/plugins/filters/geoip
```
    input {
      heartbeat {
        message => '{"host": "55.159.212.43", "app": "barfoo", "amount": 82.95}'
        interval => 10
      }
    }
    filter {
      json {
        source => "message"
      }
      geoip {
        source => "[host]"
        target => "[server]"
      }
    }
    output {
      stdout { codec => rubydebug }
    }
```

### CODE REVIEW

**Comment by yaauie** [comment]
🤔 

I am a little confused what the use-case is for standing up a "mock" GeoIP database provider serving static assets, when it would be simpler to allow a user to turn off the auto-updates. I understand that the license to use the free assets expires, but it feels like we are just pushing that problem down the road and adding complexity at the same time.

Can we instead allow a user to configure the provider as "airgapped", which _requires_ them to provide their own MMDB for which they accept responsibility for complying with terms and avoids doing any updates? Is there a simpler path forward?

***

Because the geoip provider is a part of logstash core, and can only power a single set of databases, I do not see a way to have a per-instance configuration work. Beyond just making the config unweildy, it would also open up edge-cases like a user wanting for some reason to source different databases for different instances of the plugin. For these reasons, I see a global config in Logstash core being the right place for this.

File: x-pack/lib/filters/geoip/database_manager.rb [hunk 1]
```diff
@@ -302,6 +302,11 @@ def metric=(metric)
     @metric = metric
   end
 
+  def metric
+    # Fallback when testing plugin and no metric collector are correctly configured.
```
Comment (kaisecheng) [inline x-pack/lib/filters/geoip/database_manager.rb:306]: fix the failing test cases in plugin

File: x-pack/lib/filters/geoip/database_manager.rb [hunk 2]
```diff
@@ -302,6 +302,11 @@ def metric=(metric)
...
+    @metric.nil? ? LogStash::Instrument::NamespacedNullMetric.new: @metric
```
Comment (kares) [inline x-pack/lib/filters/geoip/database_manager.rb:307]: if this is just for tests, could we mock or inject the NullMetric?
it looks weird that we instantiate a new one (which is fine for tests), 
should be also fine setting up a `@metric ||= LogStash::Instrument::NamespacedNullMetric.new` during initialize

**Review by kares — COMMENTED** [review]
code wise :+1: - easy to understand and seems to have all the parts.
one suggestion to improve `metric` 'mocking' in tests.

:warning: did not test this locally with a geoip server

File: x-pack/lib/filters/geoip/database_manager.rb (hunk 2 above)
Comment (kaisecheng) [reply x-pack/lib/filters/geoip/database_manager.rb]: It is possible to mock in plugin test but would need to mock and prepare in a way that may look unnatural. So doing in `#metric` make things simpler.  I took the suggestion `@metric ||= `, thank you.

**Review by jsvd — APPROVED** [review]
tested and LGTM.

Unfortunately faraday seems to set an exception's cause and message to the same value:

```
irb(main):018:0> begin; Faraday.new.get("http://localhost:8081"); rescue => e; puts e.message.inspect; puts e.cause.inspect; end
"Failed to open TCP connection to localhost:8081 (Connection refused - connect(2) for \"localhost\" port 8081)"
#<Errno::ECONNREFUSED: Failed to open TCP connection to localhost:8081 (Connection refused - connect(2) for "localhost" port 8081)>
```

This results in error log entries like this in Logstash:

```
[2021-09-23T10:49:53,260][ERROR][logstash.filters.geoip.databasemanager] Connection refused - Failed to open TCP connection to localhost:8080 (Connection refused - connect(2) for "localhost" port 8080) {:cause=>#<Errno::ECONNREFUSED: Connection refused - Failed to open TCP connection to localhost:8080 (Connection refused - connect(2) for "localhost" port 8080)>}
```

Or if the endpoint response is bogus:

```
[2021-09-23T10:50:52,150][ERROR][logstash.filters.geoip.databasemanager] wrong status line: "werwer" {:cause=>#<Net::HTTPBadResponse: wrong status line: "werwer">}
```

Maybe in the future we can remove faraday, but we can merge this as is.

**Comment by mtudisco49** [comment]
hi, i was running logstash 7.5, and after upgrade to 7.16.2 i started experiencing a strange behaivour.
i use geoip with database parameter 
```
                 geoip { source => "[checkpoint][origin]"
                         target => "[checkpoint][origin-GEOIP]"
                         database => "/etc/logstash/config/GeoLite2-City.mmdb"
                       }
```

We had a cron job that once a week connects to maxmind and downloads the lastest database and copies it to /etc/logstash/config/GeoLite2-City.mmdb

After upgrade, it seems that when it copies the database, the pipeline stops working with an error and after a while i start getting errors of max open files:
```
[2022-02-09T08:00:08,710][ERROR][logstash.javapipeline    ][checkpoint] Pipeline worker error, the pipeline will be stopped {:pipeline_id=>"checkpoint", :erro
r=>"Index 83 out of bounds for length 16", :exception=>Java::JavaLang::ArrayIndexOutOfBoundsException, :backtrace=>["com.maxmind.db.Decoder$Type.get(com/maxmi
nd/db/Decoder.java:52)", "com.maxmind.db.Decoder.decode(com/maxmind/db/Decoder.java:128)", "com.maxmind.db.Decoder.decode(com/maxmind/db/Decoder.java:87)", "c
om.maxmind.db.Reader.resolveDataPointer(com/maxmind/db/Reader.java:252)", "com.maxmind.db.Reader.get(com/maxmind/db/Reader.java:150)", "com.maxmind.geoip2.Dat
abaseReader.get(com/maxmind/geoip2/DatabaseReader.java:151)", "com.maxmind.geoip2.DatabaseReader.city(com/maxmind/geoip2/DatabaseReader.java:202)", "org.logst
ash.filters.geoip.GeoIPFilter.retrieveCityGeoData(org/logstash/filters/geoip/GeoIPFilter.java:234)", "org.logstash.filters.geoip.GeoIPFilter.handleEvent(org/l
ogstash/filters/geoip/GeoIPFilter.java:177)", "jdk.internal.reflect.GeneratedMethodAccessor95.invoke(jdk/internal/reflect/GeneratedMethodAccessor95)", "jdk.in
ternal.reflect.DelegatingMethodAccessorImpl.invoke(jdk/internal/reflect/DelegatingMethodAccessorImpl.java:43)", "java.lang.reflect.Method.invoke(java/lang/ref
lect/Method.java:566)", "org.jruby.javasupport.JavaMethod.invokeDirectWithExceptionHandling(org/jruby/javasupport/JavaMethod.java:456)", "org.jruby.javasuppor
t.JavaMethod.invokeDirect(org/jruby/javasupport/JavaMethod.java:317)", "usr.share.logstash.vendor.bundle.jruby.$2_dot_5_dot_0.gems.logstash_minus_filter_minus
_geoip_minus_7_dot_2_dot_8_minus_java.lib.logstash.filters.geoip.filter(/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-filter-geoip-7.2.8-java/li
b/logstash/filters/geoip.rb:117)", "usr.share.logstash.logstash_minus_core.lib.logstash.filters.base.do_filter(/usr/share/logstash/logstash-core/lib/logstash/
filters/base.rb:159)", "usr.share.logstash.logstash_minus_core.lib.logstash.filters.base.multi_filter(/usr/share/logstash/logstash-core/lib/logstash/filters/b
ase.rb:178)", "org.jruby.RubyArray.each(org/jruby/RubyArray.java:1821)", "usr.share.logstash.logstash_minus_core.lib.logstash.filters.base.multi_filter(/usr/s
hare/logstash/logstash-core/lib/logstash/filters/base.rb:175)", "org.logstash.config.ir.compiler.FilterDelegatorExt.doMultiFilter(org/logstash/config/ir/compi
ler/FilterDelegatorExt.java:127)", "org.logstash.config.ir.compiler.AbstractFilterDelegatorExt.multi_filter(org/logstash/config/ir/compiler/AbstractFilterDele
gatorExt.java:134)", "org.logstash.generated.CompiledDataset3.compute(org/logstash/generated/CompiledDataset3)", "org.logstash.generated.CompiledDataset6.comp
ute(org/logstash/generated/CompiledDataset6)", "org.logstash.generated.CompiledDataset3.compute(org/logstash/generated/CompiledDataset3)", "org.logstash.gener
ated.CompiledDataset14.compute(org/logstash/generated/CompiledDataset14)", "org.logstash.config.ir.CompiledPipeline$CompiledUnorderedExecution.compute(org/log
stash/config/ir/CompiledPipeline.java:329)", "org.logstash.config.ir.CompiledPipeline$CompiledUnorderedExecution.compute(org/logstash/config/ir/CompiledPipeli
ne.java:323)", "org.logstash.execution.WorkerLoop.run(org/logstash/execution/WorkerLoop.java:87)", "jdk.internal.reflect.GeneratedMethodAccessor83.invoke(jdk/
internal/reflect/GeneratedMethodAccessor83)", "jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(jdk/internal/reflect/DelegatingMethodAccessorImpl.java
:43)", "java.lang.reflect.Method.invoke(java/lang/reflect/Method.java:566)", "org.jruby.javasupport.JavaMethod.invokeDirectWithExceptionHandling(org/jruby/jav
asupport/JavaMethod.java:441)", "org.jruby.javasupport.JavaMethod.invokeDirect(org/jruby/javasupport/JavaMethod.java:305)", "usr.share.logstash.logstash_minus
_core.lib.logstash.java_pipeline.start_workers(/usr/share/logstash/logstash-core/lib/logstash/java_pipeline.rb:299)", "org.jruby.RubyProc.call(org/jruby/RubyP
roc.java:318)", "java.lang.Thread.run(java/lang/Thread.java:829)"], :thread=>"#<Thread:0x302ceb6c sleep>"}
[root@hostname logstash]# systemctl start logstash
```

This is no longer a valid option to actualize the database?

**Pull request merged**